### PR TITLE
Improve UI experience for the pomodoro script

### PR DESCRIPTION
**Pull request by jnduli**
Improves the UI such that:
- I can mark tasks as done, cancelled and they'll show with color and strikethrough
- I can add tasks
- Less clutter in output

### CODE REVIEW

**Comment by BonfaceKilz** [comment]
John Nduli Kilyungi ***@***.***>
aliandika:

> Improves the UI such that:
> - I can mark tasks as done, cancelled and they&#39;ll show with color and strikethrough
> - I can add tasks
> - Less clutter in output
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/jnduli/pomodoro/pull/32
>
> -- Commit Summary --
>
>   * plans for better output
>   * better output plan
>   * demo list separated list and color output of done tasks
>   * add support for abandoned tasks
>   * clean up interactive tasks
>   * add interactive output to pomodoro main script
>   * bug fixes for incorrect output
>   * rough fix of some bugs in interactive tests
>   * attempt to fix line breaks
>   * experiments
>   * fix pomodoro line outputs using clearline
>   * minor bug fix
>   * improvements to pomodoro output
>
> -- File Changes --
>
>     A better_output.md (44)
>     A interactive_test.sh (143)
>     M pomodoro.sh (157)
>
> -- Patch Links --
>

Nice! +1 for this.  I should now give this a shot.
But before I do, it would be a great first package
for Ubunifu which I've ignored for months!

-- 
(Life is like a pencil that will surely run out,
    but will leave the beautiful writing of life.)
(D4F09EB110177E03C28E2FE1F5BBAE1E0392253F
    (hkp://keys.openpgp.org))

File: better_output.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
+# Better Outputs
```
Comment (jnduli) [inline better_output.md]: drop this file and update the documentation. Create an ASCIInema showing the usage of the script

File: interactive_test.sh [hunk 2]
```diff
@@ -0,0 +1,143 @@
+#!/bin/bash
```
Comment (jnduli) [inline interactive_test.sh]: drop this file, was used to demo the script before attempting the change the main script

File: pomodoro.sh [hunk 3]
```diff
@@ -31,6 +31,17 @@ if [ -f "$CONFIG_FILE" ]; then
...
 fi
 
+# Global variables
+TODO=()
+declare -A COMPLETED
+declare -A ABANDONED
+CURRENT_TASK="work" # temporary value to help determine child tasks, TODO: Drop this
+
+# Colors used in display
+# Set the color variable
```
Comment (jnduli) [inline pomodoro.sh]: unnecessary comment

File: pomodoro.sh [hunk 4]
```diff
@@ -73,39 +85,61 @@ clear_line () {
...
         secs_to_count_down=$((($2+1)*$1*SECS_IN_MINUTE))
     fi
     while (( SECONDS <= secs_to_count_down )); do    # Loop until interval has elapsed.
         minutes=$((SECONDS/SECS_IN_MINUTE))
-        if [[ $printed_minutes != "$minutes" ]]; then # updates screen after every minute, preventing stuttering
+        if [[ $changed == 't' || $printed_minutes != "$minutes" ]]; then # updates screen after every minute, preventing stuttering
             printed_minutes=$minutes
-            clear_line
-            echo -e "\t\tTime spent $printed_minutes minutes"
+            clear_line $(( pomodoro_lines + 1 ))
```
Comment (jnduli) [inline pomodoro.sh:114]: why +1 here? Add a comment explanation

File: pomodoro.sh [hunk 5]
```diff
@@ -73,39 +85,61 @@ clear_line () {
...
-    echo -e "\t\tTime spent $printed_minutes minutes"
     SECONDS=0 
     if [ -n "$2" ]; then
         SECONDS=$(($1*SECS_IN_MINUTE*$2))
         secs_to_count_down=$((($2+1)*$1*SECS_IN_MINUTE))
     fi
     while (( SECONDS <= secs_to_count_down )); do    # Loop until interval has elapsed.
         minutes=$((SECONDS/SECS_IN_MINUTE))
-        if [[ $printed_minutes != "$minutes" ]]; then # updates screen after every minute, preventing stuttering
+        if [[ $changed == 't' || $printed_minutes != "$minutes" ]]; then # updates screen after every minute, preventing stuttering
```
Comment (jnduli) [inline pomodoro.sh]: cast `$changed` to lower

File: pomodoro.sh [hunk 6]
```diff
@@ -73,39 +85,61 @@ clear_line () {
...
             printed_minutes=$minutes
-            clear_line
-            echo -e "\t\tTime spent $printed_minutes minutes"
+            clear_line $(( pomodoro_lines + 1 ))
+            pomodoro=$(refresh_current_pomodoro_output)
```
Comment (jnduli) [inline pomodoro.sh]: move pomodoro call to within the `if work` block

File: pomodoro.sh [hunk 7]
```diff
@@ -172,6 +206,82 @@ EOF
     eval "$1=$should_continue" # set first parameter to have the return type
 }
 
+############################################
+# TODO: adding better output 
+# ###########################################
+#
+STATIC_OUTPUT=""
```
Comment (jnduli) [inline pomodoro.sh]: unused, remove from script

File: pomodoro.sh [hunk 8]
```diff
@@ -172,6 +206,82 @@ EOF
...
+DYNAMIC_OUTPUT=""
+
+refresh_screen() {
```
Comment (jnduli) [inline pomodoro.sh]: unused function, remove from script

File: pomodoro.sh [hunk 9]
```diff
@@ -172,6 +206,82 @@ EOF
...
+    # so I'll maintain a global list of what needs to change and add to it as pomodoros complete
+    # I'll then use this to refresh the whole screen and call this on each action I get and work on
+    tput clear # clear the whole screen
+    printf "%b" "$STATIC_OUTPUT"
+    printf "%b" "$DYNAMIC_OUTPUT"
+}
+
+
+refresh_current_pomodoro_output () {
+    # set -x
```
Comment (jnduli) [inline pomodoro.sh]: remove debugging information and add comment in README on how to debug bash functions

File: pomodoro.sh [hunk 10]
```diff
@@ -172,6 +206,82 @@ EOF
...
+            local_output="$i: $strikethrough${TODO[i]}$clear"
+        else
+            local_output="$i: ${TODO[i]}"
+        fi
+        local non_color_output="$non_color_last_line $i: ${TODO[i]}, "
+        if [[ ${#non_color_output}+$prefix_chars+4 -gt $COLUMNS ]]; then
+            output+=("$local_output, ")
+            non_color_last_line="$i: ${TODO[i]}, "
+        else
+            local potential_output="${output[-1]} $local_output, "
```
Comment (jnduli) [inline pomodoro.sh]: no need for `potential_output` variable

File: pomodoro.sh [hunk 11]
```diff
@@ -172,6 +206,82 @@ EOF
...
+add_to_list() {
+    read -r -p "Additional Tasks: " tasks 
+    readarray -td', ' temp_arr <<< "$tasks" # comma separated input
+    TODO=(${TODO[@]} ${temp_arr[@]})
+    clear_line 1
+}
+
+complete_task() { # rename to complete task
+    read -r -p "Tasks no: " task_no
+    COMPLETED["$task_no"]="this"
```
Comment (jnduli) [inline pomodoro.sh]: change this to done

File: pomodoro.sh [hunk 12]
```diff
@@ -172,6 +206,82 @@ EOF
...
+    clear_line 1
+}
+
+cancel_task() {
+    read -r -p "Tasks no: " task_no
+    ABANDONED["$task_no"]="this"
```
Comment (jnduli) [inline pomodoro.sh]: change this to abandoned

File: pomodoro.sh [hunk 13]
```diff
@@ -182,15 +292,28 @@ single_pomodoro_run () {
     echo "Pomodoro $1"
     read -r -p 'Plan: ' work
     clear_line
-    echo -e '\tPlan: ' "$work"
+    echo -e "\tPlan: $work"
+    # TODO: add relevant comment from this shttps://unix.stackexchange.com/a/519917
```
Comment (jnduli) [inline pomodoro.sh]: address the TODO here

File: pomodoro.sh [hunk 14]
```diff
@@ -286,6 +410,7 @@ main () {
     options "$@"
     rename_window_in_tmux
     echo "Starting pomodoro, work=$WORK and rest=$REST minutes"
+    STATIC_OUTPUT=$(echo "Starting pomodoro, work=$WORK and rest=$REST minutes")
```
Comment (jnduli) [inline pomodoro.sh]: drop STATIC_OUTPUT

File: pomodoro.sh [hunk 15]
```diff
@@ -182,15 +292,28 @@ single_pomodoro_run () {
...
-    echo -e "\tStarting work:"
-    work_or_rest $WORK
+    echo -e "\tStarting work:\n"
+    CURRENT_TASK="work"
+    work_or_rest "$WORK"
+
+    # add to STATIC CONTENT before we reset values
+    # TODO: figure out if I want to add rest to this static output too
+    pomodoro_content=$(refresh_current_pomodoro_output)
+    STATIC_OUTPUT=$(printf "%b\nPomodoro %d\n%b" "$STATIC_OUTPUT" "$1" "$pomodoro_content")
```
Comment (jnduli) [inline pomodoro.sh]: drop STATIC_OUTPUT

File: interactive_test.sh (hunk 2 above)
Comment (jnduli) [reply interactive_test.sh]: done

File: pomodoro.sh (hunk 3 above)
Comment (jnduli) [reply pomodoro.sh]: done

**Pull request merged**